### DISCUSSION TITLE
Apex Custom PMD Rule for string concatenation without + operator.

### DISCUSSION THREAD

**Discussion by aklmmap** [pmd/pmd#3078] [Q&A]
We are trying to develop a apex custom PMD rule.
Rule: String concatenation shouldnt be done with '+' operator.

We have developed the following apex pmd code in java. But it looks it triggering the violations for any statements that contains '+'. 
the '+' can be inside a string and the code still counts this as violation. It counts arithmetic addition (a=b+c) also as an violation. 

Can anyone please help me out here. 

Following is the code:

```java
public class InefficientStringOperationRule extends AbstractApexRule {

@Override
public Object visit(ASTExpressionStatement node, Object data) { 
        //to check string concatination
        try {
            String str = node.getNode().toString(); 
            if( str.contains("+") && !str.contains("System.debug") ) {
                    if(!str.contains("system.debug") )
                        addViolation(data, node);
            }
        }catch(java.util.NoSuchElementException e) {
            //Re throwing it does not make sense, nothing to do. Even puttting in logs adds no value.
        }

        return data;
}
}
```

![concatenation](https://user-images.githubusercontent.com/77477451/105456388-1536ad80-5cab-11eb-976b-8885b8cdc890.png)

**Comment by adangel** [comment]
I'd suggest to use the Designer tool for developing rules: https://pmd.github.io/latest/pmd_userdocs_extending_designer_reference.html

Looking for "+" operator, you can visit "BinaryExpression" nodes with "+" as the operator. Then you don't need this "str.contains" search, which finds "+" wherever it occurs.

You could then check the children of this binary expression. In case of string literals, it's easy to determine the type. But for variable references, you'd probably need lookup the variable declaration yourself and check the type there.

Here's an example apex code for this rule:

 ```
public class Foo {
    public bar() {
        Integer i = 1;
        i = i + 1; // this should not be detected, valid usage of '+'
        String s = 'A' + 'b' + 'c'; // inefficient
        String t = s + s; // inefficient with variables
        doSomething('a' + 'b'); // inefficient, as method call parameter
        doSomething(s + t); // inefficient with vars as method call parameter
    }
}
```

This is, how it looks like in the designer:
![grafik](https://user-images.githubusercontent.com/1573684/105486239-7ad86900-5cae-11eb-84b8-1393a559fafb.png)

As you can see, half of the rule can be simply done with the following XPath expression:

```
//BinaryExpression[@Operator='+'][LiteralExpression[@LiteralType = 'STRING']]
```

This will find string concatenation, where at least one literal string is involved. It doesn't find concats with only variables.
